src/tests/Microsoft.PowerFx.Interpreter.Tests/CustomFunctions.cs

```diff
@@ -247,6 +250,120 @@ public static RecordValue Execute()
+                var val = FormulaValue.NewRecordFromFields(recordType, new NamedValue("num1", FormulaValue.New(1)));
+                return val;
+            }
+        }
+
+        [Fact]
+        public void CustomFunctionRecordTest()
+        {
+            var oneFieldRecord = RecordType.Empty().Add(new NamedFormulaType("x", FormulaType.Number));
+            var record = oneFieldRecord.Add(new NamedFormulaType("y", FormulaType.Number));
+            var coerceRecord = RecordType.Empty().Add(new NamedFormulaType("x", FormulaType.String)).Add(new NamedFormulaType("y", FormulaType.String));
+            var nestedRecord = RecordType.Empty().Add(new NamedFormulaType("x", RecordType.Empty().Add(new NamedFormulaType("x", FormulaType.Number))));
+            var extraFieldRecord = record.Add(new NamedFormulaType("a", FormulaType.Number));
+            var notExistRecord = RecordType.Empty().Add(new NamedFormulaType("a", FormulaType.Number));
+
+            CustomFunctionRecord(new TestRecordsCustomFunction(record, record), "RecordsTest", "({x: 2, y: 5})", true);
+            CustomFunctionRecord(new TestRecordsCustomFunction(record, oneFieldRecord), "RecordsTest", "({x: 2})", true);
+            CustomFunctionRecord(new TestRecordsCustomFunction(record, coerceRecord), "RecordsTest", "({x: \"2\", y: \"5\"})", true);            
+            CustomFunctionRecord(new TestRecordsCustomFunction(record, extraFieldRecord), "RecordsTest", "({x: 2, y: 5, a: 8})", false);
+            CustomFunctionRecord(new TestRecordsCustomFunction(record, notExistRecord), "RecordsTest", "({a: 8})", false);            
+            CustomFunctionRecord(new TestRecordsCustomFunction(nestedRecord, nestedRecord), "RecordsTest", "({x: {x : 2}})", true);           
+            CustomFunctionRecord(new TestRecordsCustomFunction(record, nestedRecord), "RecordsTest", "({x: {x : 2}})", false);
+            
+            CustomFunctionRecord(new TestNonRecordCustomFunction(), "NonRecordTest", "({x: 2})", false);
+        }
+
+        private void CustomFunctionRecord(ReflectionFunction reflectionFunction, string functionName, string inputRecord,  bool isSuccess)
+        {
+            var config = new PowerFxConfig();
+            config.AddFunction(reflectionFunction);
+            
+            var engine = new RecalcEngine(config);
+            var func = engine.GetFunctionsByName(functionName);
+            Assert.NotNull(func);
+
+            var check = engine.Check(functionName + inputRecord);
+
+            FormulaValue result = null; 
+            ConfiguredTaskAwaitable<FormulaValue> resultAsync;
+            string errorMsg = string.Empty;
+
+            try
+            {
+                result = engine.Eval(functionName + inputRecord);
+
+                using var cts = new CancellationTokenSource();
+                resultAsync = engine.EvalAsync(functionName + inputRecord, cts.Token).ConfigureAwait(false);
```

Comment:
>EvalAsync [](http://example.com/codeflow?start=37&length=9)

I think we can skip EvalAsync part since if Check().IsSuccess() is false we always throw.

Reply:
thinking back I think its fine, since you are also testing non-error scenario.